pkg/volume/scaleio/sio_volume.go

```diff
@@ -0,0 +1,390 @@
+	return nil
+}
+
+func (v *sioVolume) GetAttributes() volume.Attributes {
+	return volume.Attributes{
+		ReadOnly:        v.readOnly,
+		Managed:         !v.readOnly,
+		SupportsSELinux: true,
+	}
+}
+
+// **********************
+// volume.Unmounter Impl
+// *********************
+var _ volume.Unmounter = &sioVolume{}
+
+// TearDownAt unmounts the bind mount
+func (v *sioVolume) TearDown() error {
+	return v.TearDownAt(v.GetPath())
+}
+
+// TearDown unmounts  and remove the volume
+func (v *sioVolume) TearDownAt(dir string) error {
+	v.plugin.volumeMtx.LockKey(v.volSpecName)
+	defer v.plugin.volumeMtx.UnlockKey(v.volSpecName)
+
+	glog.V(4).Info(log("attempting to unmount %s", dir))
+	if err := util.UnmountPath(dir, v.plugin.mounter); err != nil {
+		glog.Error(log("teardown failed while unmounting dir %s: %v ", dir, err))
+		return err
+	}
+	glog.V(4).Info(log("dir %s unmounted successfully", dir))
+
+	// detach/unmap
+	glog.V(4).Info(log("teardown is attempting to detach/unmap volume for %s", v.volSpecName))
+	if err := v.resetSioMgr(); err != nil {
+		glog.Error(log("teardown failed, unable to reset scalio mgr: %v", err))
+	}
+	volName := v.volName
+	if err := v.sioMgr.DetachVolume(volName); err != nil {
+		glog.Error(log("teardown failed detaching volume %s:  %v", volName, err))
+		return err
+	}
+
+	glog.V(4).Infof(log("teardown of volume %v detached successfully", volName))
+	return nil
+}
+
+// ********************
+// volume.Deleter Impl
+// ********************
+var _ volume.Deleter = &sioVolume{}
+
+func (v *sioVolume) Delete() error {
+	glog.V(4).Info(log("deleting pvc %s", v.volSpecName))
+	if err := v.resetSioMgr(); err != nil {
+		glog.Error(log("delete failed, unable to reset scalio mgr: %v", err))
+	}
+	err := v.sioMgr.DeleteVolume(v.volName)
+	if err != nil {
+		glog.Error(log("failed to delete volume %s: %v", v.volName, err))
+		return err
+	}
+
+	glog.V(4).Info(log("successfully deleted pvc %s", v.volSpecName))
+	return nil
+}
+
+// ************************
+// volume.Provisioner Impl
+// ************************
+var _ volume.Provisioner = &sioVolume{}
+
+func (v *sioVolume) Provision() (*api.PersistentVolume, error) {
+	glog.V(4).Info(log("attempting to dynamically provision pvc %v", v.options.PVName))
+
+	// setup volume attrributes
+	name := v.generateVolName()
+	capacity := v.options.PVC.Spec.Resources.Requests[api.ResourceName(api.ResourceStorage)]
+	volSizeBytes := capacity.Value()
+	volSizeGB := int64(volume.RoundUpSize(volSizeBytes, 1024*1024*1024))
+
+	if err := v.setSioMgrFromConfig(); err != nil {
+		glog.Error(log("provision failed while setting up sio mgr: %v", err))
+		return nil, err
+	}
+
+	vol, err := v.sioMgr.CreateVolume(name, volSizeGB)
+	if err != nil {
+		glog.Error(log("provision failed while creating volume: %v", err))
+		return nil, err
+	}
+
+	namespace, _ := splitVolumeName(v.options.PVName)
```

Comment:
@rootfs  The design approach in this plugin is to prefix `volume.Name` with the namespace value using a separator char (e.g. `mynamespace%myvlolname`).  This workaround is because when the PV controller is doing reconcilations by calling `NewUnmounter(volName string, podUID types.UID)`, the plugin code does not have access to a full spec in order to get the `namespace` value.  Adding the prefix to the volName is a way to access the namespace regardless of the controller cycle.